### PR TITLE
Auto: implement issue #582

### LINKED ISSUE #582: Add lead_score columns to CustomerModel and scoring Pydantic schemas
Subtask of #49

Add `score: int (0-100, nullable)`, `tier: str (nullable)`, `score_factors: dict (nullable)`, `top_factors: list (nullable)`, `recommendations: list (nullable)` columns to the CustomerModel ORM. Add `ScoreRequest`, `ScoreResponse` Pydantic schemas matching the issue's output shape. Create and run an alembic migration. Touch src/db/models/ and alembic/versions/.

Dev-Plan: docs/dev-plan/10-customers/0582-add-lead-score-columns-to-customermodel-and-scoring-pydantic.md

### PR DESCRIPTION
**Pull request by yihua-zhuo**
Closes #582

Plan: `.plans/issue-582.md`

This PR was auto-implemented by Claude Code in three stages:

1. **Plan** — Claude wrote a detailed implementation plan to `.plans/issue-582.md`.
2. **Implement** — Claude executed the plan, ran `ruff check src/` + the unit and integration test suites, and iterated until all three were green.
3. **Self-review** — Claude reviewed its own diff against `codereview.md` (the project review checklist) and fixed any rule violations found, re-running the test suite if any changes were made.

PR-time CI (`Dev Agent CI`) re-verifies independently before this PR is mergeable.

Please review before merging.

## Release Notes

* **New Features**
  * Added lead scoring functionality to track customer scores, tiers, and recommendations
  * Introduced new API schemas for lead scoring requests and responses

* **Chores**
  * Updated database schema to support lead scoring fields
  * Added migration to resolve multiple database branches

* **Tests**
  * Added comprehensive unit test coverage for scoring schemas and validation

[Review Change Stack](https://app.coderabbit.ai/change-stack/yihua-zhuo/agent-job/pull/696?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: alembic/versions/e8ad92ca62b3_add_lead_score_columns.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+from __future__ import annotations
+
+import sqlalchemy as sa
+from sqlalchemy.dialects import postgresql
+
+from alembic import op
+
+# revision identifiers, used by Alembic.
+revision: str = 'e8ad92ca62b3'
+down_revision: str | None = 'eea413253aa1'
```
Comment (yihua-zhuo) [inline alembic/versions/e8ad92ca62b3_add_lead_score_columns.py:17]: **CRITICAL** (critical)

**Issue:** down_revision points to 'eea413253aa1' which is a merge revision (has two parents: '9d8e7f6a5b3c' and 'c94d682d4b03'). Alembic requires a single linear down_revision; pointing at a merge head is fragile — if the merge revision is ever removed or the migration chain restructured, this migration becomes unreachable.

**Suggestion:** Change down_revision to the last linear migration before the merge, e.g. down_revision = 'c94d682d4b03' (or '9d8e7f6a5b3c') so the chain is linear. The merge revision's purpose (unifying two branch heads) is valid but should not be an ancestor of a later migration's linear chain.

File: .plans/issue-582.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+## Implementation Steps
+1. **Add ORM columns to `CustomerModel`** (`src/db/models/customer.py`):
+   - Import `Integer` and `JSON` alongside existing imports
+   - Add `score: Mapped[int | None] = mapped_column(Integer, nullable=True)` after `recycle_history`
+   - Add `tier: Mapped[str | None] = mapped_column(String(50), nullable=True)` after `score`
+   - Add `score_factors: Mapped[dict | None] = mapped_column(JSON, nullable=True)` after `tier`
+   - Add `top_factors: Mapped[list | None] = mapped_column(JSON, nullable=True)` after `score_factors`
+   - Add `recommendations: Mapped[list | None] = mapped_column(JSON, nullable=True)` after `top_factors`
+   - Update `to_dict()` to include all five new fields with appropriate None defaults (`self.score`, `self.tier`, `self.score_factors or {}`, `self.top_factors or []`, `self.recommendations or []`)
```
Comment (yihua-zhuo) [inline .plans/issue-582.md:22]: **WARNING** (suggestion)

**Issue:** Plan references 'src/db/models/customer.py' correctly for ORM columns, but there are two customer model files (src/models/customer.py is a Pydantic dataclass; src/db/models/customer.py is the ORM); path is correct here but the naming overlap could confuse contributors.

**Suggestion:** Add a clarifying comment noting the ORM file is src/db/models/customer.py to distinguish from src/models/customer.py

File: .plans/issue-582.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
+Now I have everything I need. Here's the implementation plan:
+
+# Implementation Plan — Issue #582
+
+## Goal
+Add five nullable lead-score columns (`score`, `tier`, `score_factors`, `top_factors`, `recommendations`) to `CustomerModel` and create `ScoreRequest`/`ScoreResponse` Pydantic schemas. An alembic migration will apply the schema change to the database.
+
+## Affected Files
+- `src/db/models/customer.py` — add five ORM columns and update `to_dict()`
```
Comment (yihua-zhuo) [inline .plans/issue-582.md:9]: **WARNING** (suggestion)

**Issue:** Plan step 4 references 'tests/unit/domain_handlers/customers.py' but the actual file is at tests/unit/domain_handlers/customers.py (plural 'handlers'); step 5 references 'tests/unit/test_customer_model.py' which exists but tests the Pydantic Customer, not the ORM CustomerModel.

**Suggestion:** Clarify file paths: step 4 → tests/unit/domain_handlers/customers.py (plural), step 5 → the unit test for ORM is not yet present; suggest creating tests/unit/test_customer_orm.py or testing via the existing test_customer_model.py but note it tests the Pydantic model not the ORM

File: .plans/issue-582.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+   - Add `score_factors: Mapped[dict | None] = mapped_column(JSON, nullable=True)` after `tier`
+   - Add `top_factors: Mapped[list | None] = mapped_column(JSON, nullable=True)` after `score_factors`
+   - Add `recommendations: Mapped[list | None] = mapped_column(JSON, nullable=True)` after `top_factors`
+   - Update `to_dict()` to include all five new fields with appropriate None defaults (`self.score`, `self.tier`, `self.score_factors or {}`, `self.top_factors or []`, `self.recommendations or []`)
+
+2. **Create `src/models/score.py`** with two Pydantic schemas:
+   - `ScoreRequest(BaseModel)`: fields matching what the scoring engine needs as input (e.g. `customer_id`, `tenant_id`, optional raw attributes). Fields and types should match the existing `SmartCategorizationService.score_lead()` input shape.
+   - `ScoreResponse(BaseModel)`: fields `score: int | None`, `tier: str | None`, `score_factors: dict | None`, `top_factors: list | None`, `recommendations: list | None`, plus a `to_dict()` method. Both schemas should follow the same `Annotated[..., Field(...)]` pattern used in `CustomerCreateDTO`.
+
```
Comment (yihua-zhuo) [inline .plans/issue-582.md:27]: **INFO** (suggestion)

**Issue:** Plan step 2 says ScoreResponse should follow the Annotated[..., Field(...)] pattern from CustomerCreateDTO, but src/models/score.py uses plain field annotations without Field(); this may be an acceptable simplification but is a deviation from the stated plan.

**Suggestion:** Align ScoreResponse field definitions with the Annotated[..., Field(...)] pattern if stricter validation is desired, or update the plan to reflect the simpler plain-annotation approach actually used

File: alembic/versions/eea413253aa1_merge_heads.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+branch_labels: list[str] | None = None
+depends_on: list[str] | None = None
+
+
+def upgrade() -> None:
+    pass
+
+
+def downgrade() -> None:
+    pass
```
Comment (yihua-zhuo) [inline alembic/versions/eea413253aa1_merge_heads.py:22]: **WARNING** (suggestion)

**Issue:** The merge revision eea413253aa1 has two parents (9d8e7f6a5b3c and c94d682d4b03) but upgrade()/downgrade() are both pass. This is structurally valid for a merge but leaves no artifact of the merge in the migration history — alembic history will show a gap between the two branches until the merge revision stamps the history.

**Suggestion:** Verify that alembic is configured with 'branch_labels' or a merge workflow that marks the merge revision as a permanent branch point. If this is intentional (i.e., the two branches were developed in parallel and this revision simply unifies them without schema changes), add a docstring clarifying the merge purpose so future developers do not try to 'fix' the empty functions.

File: src/db/models/customer.py [hunk 6]
```diff
@@ -51,6 +56,11 @@ def to_dict(self) -> dict:
             "assigned_at": self.assigned_at.isoformat() if self.assigned_at else None,
             "recycle_count": self.recycle_count or 0,
             "recycle_history": self.recycle_history or [],
+            "score": self.score,
```
Comment (yihua-zhuo) [inline src/db/models/customer.py:59]: **WARNING** (suggestion)

**Issue:** Plan step 1 specifies score_factors defaults to {} (or empty dict), top_factors to [], and recommendations to [] via or-chains. However, returning an empty dict/list for null DB values may be indistinguishable from a deliberately set empty object, and creates inconsistency with email/phone which return None for null values.

**Suggestion:** If the intent is to distinguish 'not scored yet' (None) from 'scored with no factors' ({}, []), this behavior is correct. If None is preferred for all unset fields for consistency, remove the or-chains and return self.score_factors, self.top_factors, self.recommendations directly. Document the chosen convention in a class-level docstring.

File: src/models/score.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+            "engaged_actions": self.engaged_actions,
+        }
+
+
+class ScoreResponse(BaseModel):
+    """Output schema returned after scoring a lead."""
+
+    score: int | None = None
+    tier: str | None = None
+    score_factors: dict | None = None
```
Comment (yihua-zhuo) [inline src/models/score.py:41]: **WARNING** (suggestion)

**Issue:** ScoreResponse.score_factors uses bare 'dict | None' — unparameterized generic conceals key/value types

**Suggestion:** Use Annotated[dict[str, Any], Field(default=None)] or dict[str, Any] | None

File: src/models/score.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    top_factors: list | None = None
```
Comment (yihua-zhuo) [inline src/models/score.py:42]: **WARNING** (suggestion)

**Issue:** ScoreResponse.top_factors and recommendations use bare 'list | None' — unparameterized generic conceals item type

**Suggestion:** Use list[str] | None for both fields

File: src/models/score.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    model_config = {"str_strip_whitespace": True}
+
+    customer_id: Annotated[int, Field(gt=0)]
+    tenant_id: Annotated[int, Field(gt=0)]
+    source: Annotated[str, Field(min_length=1, max_length=50)]
+    company_size: Annotated[int, Field(ge=0)]
+    title: Annotated[str, Field(min_length=1, max_length=255)]
+    engaged_actions: list[str] = Field(default_factory=list)
+
+    def to_dict(self) -> dict[str, Any]:
```
Comment (yihua-zhuo) [inline src/models/score.py:24]: **INFO** (suggestion)

**Issue:** to_dict() is defined on both ScoreRequest and ScoreResponse; Pydantic v2 provides model_dump() which may render these redundant

**Suggestion:** Confirm these are consumed by a router layer that calls .to_dict() explicitly, otherwise delete both methods and use .model_dump() instead

**Comment by yihua-zhuo** [comment]
## Claude Code Review (inline)

**Status:** `fail`
**Inline comments posted:** 9

**Summary:** Reviewed 7 file(s) between origin/master and origin/auto/issue-582.

### Issues without a usable file/line
- `tests/unit/test_score_schemas.py`: No issues found — tests are well-structured with clear intent, good field-level assertions, and comprehensive edge-case coverage

### Issues that could not be posted inline
`tests/unit/domain_handlers/customers.py` L80: 
**WARNING** (suggestion)

**Issue:** COUNT handler returns hardcoded 3/7 without explanation; these magic numbers have no anchor in fixture logic and could mislead future readers

**Suggestion:** Extract to named constants or derive from len([c for c in state.customers.values() if c['tenant_id'] == tenant_id]) for realism — _HTTP 422: {"message":"Validation Failed","errors":[{"resource":"PullRequestReviewComment","code":"custom","field":"pull_request_review_thread.line","message":"could not be resolved"}],"documentation_url":"https://docs.github.com/rest/pulls/comments#create-a-review-comment-for-a-pull-request","status":"422"}_
`tests/unit/domain_handlers/customers.py` L96: 
**WARNING** (suggestion)

**Issue:** SELECT WHERE ID fixtures hardcode tenant_id=1 for all fallback records — cross-tenant negative tests (rule 126) cannot work with these fixtures

**Suggestion:** Make fallback fixtures parametric on an explicit tenant_id param passed to the handler, or document clearly that this handler is not suitable for cross-tenant test scenarios — _HTTP 422: {"message":"Validation Failed","errors":[{"resource":"PullRequestReviewComment","code":"custom","field":"pull_request_review_thread.line","message":"could not be resolved"}],"documentation_url":"https://docs.github.com/rest/pulls/comments#create-a-review-comment-for-a-pull-request","status":"422"}_
`tests/unit/domain_handlers/customers.py` L50: 
**WARNING** (suggestion)

**Issue:** UPDATE path for unknown customer_id >= 1 creates a phantom record without validating tenant_id scope — breaks tenant isolation (rule 134)

**Suggestion:** Return MockResult([]) for the 'not in state.customers' case instead of fabricating a record; UPDATE on a missing row should be a no-op, not an unvalidated upsert — _HTTP 422: {"message":"Validation Failed","errors":[{"resource":"PullRequestReviewComment","code":"custom","field":"pull_request_review_thread.line","message":"could not be resolved"}],"documentation_url":"https://docs.github.com/rest/pulls/comments#create-a-review-comment-for-a-pull-request","status":"422"}_
`tests/unit/domain_handlers/customers.py` L78: 
**INFO** (suggestion)

**Issue:** DELETE handler is a stub — it returns MockRow({'id': params.get('id', 1)}) without removing any record from state.customers, so tests asserting deletion side effects will pass incorrectly

**Suggestion:** Remove the record from state.customers when the delete matches, and return MockResult([]) when no matching record exists — _HTTP 422: {"message":"Validation Failed","errors":[{"resource":"PullRequestReviewComment","code":"custom","field":"pull_request_review_thread.line","message":"could not be resolved"}],"documentation_url":"https://docs.github.com/rest/pulls/comments#create-a-review-comment-for-a-pull-request","status":"422"}_
`tests/unit/domain_handlers/customers.py` L9: 
**INFO** (suggestion)

**Issue:** ORDER = 10 module-level constant; no other domain handler sets this — creates an inconsistency and implies get_handlers() is sorted by order, but the sorting is never used

**Suggestion:** Remove ORDER if unused, or document its purpose; verify conftest._load_domain_handler_modules actually calls the sort by ORDER logic — _HTTP 422: {"message":"Validation Failed","errors":[{"resource":"PullRequestReviewComment","code":"custom","field":"pull_request_review_thread.line","message":"could not be resolved"}],"documentation_url":"https://docs.github.com/rest/pulls/comments#create-a-review-comment-for-a-pull-request","status":"422"}_
